app/javascript/controllers/apex_controller.js

```diff
@@ -0,0 +1,204 @@
+import { Controller } from "@hotwired/stimulus"
+import ApexCharts from "apexcharts"
+
+const DEFAULT_OPTIONS = {
+  chart: {
+    width: "100%",
+    zoom: { enabled: false },
+    animations: {
+      speed: 1000,
+      animateGradually: {
+        delay: 50,
+      },
+      dynamicAnimation: {
+        enabled: true,
+        speed: 1000,
+      },
+    },
+  },
+  yaxis: {
+    title: { text: "Downloads" },
+  },
+}
+
+const LINE_DEFAULTS = {
+  stroke: {
+    curve: "smooth",
+    width: 2,
+  },
+  colors: [
+    "#007EB2",
+    "#FF9600",
+    "#75BBE1",
+    "#FFC107",
+    "#6F42C1",
+    "#DC3545",
+    "#198754",
+    "#D63384",
+    "#20C997",
+    "#555555",
+  ],
+  legend: {
+    show: false,
+  },
+}
+
+const BAR_DEFAULTS = {
+  plotOptions: {
+    bar: {
+      horizontal: true,
+    },
+  },
+}
+
+export default class extends Controller {
+  static values = {
+    id: String,
+    chartType: String,
+    seriesType: String,
+    seriesData: Array,
+    dateRange: Array,
+    interval: String,
+  }
+  static targets = ["chart", "episodebox", "dateview", "interval", "start"]
+
+  connect() {
+    if (this.seriesDataValue.length) {
+      const options = Object.assign({}, DEFAULT_OPTIONS)
+      const series = this.buildSeries()
+      const typeOptions = this.setChartTypeDefaults(options, this.chartTypeValue)
+      Object.assign(options, series, typeOptions)
+
+      const chart = new ApexCharts(this.chartTarget, options)
+      chart.render()
+    }
+  }
+
+  toggleSeries(event) {
+    if (event.target.checked) {
+      ApexCharts.exec(this.idValue, "showSeries", event.target.dataset.series)
+    } else {
+      ApexCharts.exec(this.idValue, "hideSeries", event.target.dataset.series)
+    }
+  }
+
+  updateDateStart(event) {
+    this.startTarget.value = event.target.value
+  }
+
+  buildSeries() {
+    if (this.seriesTypeValue === "episodeRollups" && this.dateRangeValue.length) {
+      return this.buildEpisodeRollupsSeries()
+    } else if (this.seriesTypeValue === "agents") {
+      return this.buildAgentSeries()
+    }
+  }
+
+  buildEpisodeRollupsSeries() {
+    if (this.seriesDataValue.length) {
+      return {
+        series: this.seriesDataValue.map((d) => {
+          return {
+            name: d.ep.title,
+            data: this.alignRollupsOnDateRange(d.rollups, this.dateRangeValue),
+          }
+        }),
+      }
+    } else {
+      return []
+    }
+  }
+
+  buildAgentSeries() {
+    if (this.seriesDataValue.length) {
+      return {
+        series: [
+          {
+            data: this.seriesDataValue,
+          },
+        ],
+      }
+    } else {
+      return []
+    }
+  }
+
+  alignRollupsOnDateRange(rollups, range) {
+    return range.map((date) => {
+      const utcDate = new Date(date).toUTCString()
```

Comment:
right now all times are UTC (to my knowledge), so parsing user timezones will be something to figure out later.